subxt/src/utils.rs

```diff
@@ -96,3 +96,137 @@ unsafe impl<T> Sync for PhantomDataSendSync<T> {}
+                #[derive(Clone, Debug, PartialEq, Eq)]
+                pub enum $ty {}
+                impl $ident for $ty {
+                    const FORMAT: OrderFormat = OrderFormat::$ty;
+                }
+            )*
+        };
+    }
+
+    store!(BitStore; (U8, u8), (U16, u16), (U32, u32), (U64, u64));
+    order!(BitOrder; Lsb0, Msb0);
+
+    /// Constructs a run-time format parameters based on the corresponding type-level parameters.
+    pub fn bit_format<Store: BitStore, Order: BitOrder>() -> Format {
+        Format {
+            order: Order::FORMAT,
+            store: Store::FORMAT,
+        }
+    }
+
+    /// `scale_bits::Bits` generic over the bit store (`u8`/`u16`/`u32`/`u64`) and bit order (LSB, MSB) types.
+    #[derive(Debug, Clone, PartialEq, Eq)]
+    pub struct DecodedBits<Store: BitStore, Order: BitOrder>(
+        pub Bits,
+        pub PhantomData<Store>,
+        pub PhantomData<Order>,
+    );
+
+    impl<Store: BitStore, Order: BitOrder> codec::Decode for DecodedBits<Store, Order> {
+        fn decode<I: Input>(input: &mut I) -> Result<Self, codec::Error> {
+            /// Equivalent of `BitSlice::MAX_BITS` on 32bit machine.
+            const ARCH32BIT_BITSLICE_MAX_BITS: usize = 0x1fff_ffff;
+
+            let Compact(bits) = <Compact<u32>>::decode(input)?;
+            // Otherwise it is impossible to store it on 32bit machine.
+            if bits as usize > ARCH32BIT_BITSLICE_MAX_BITS {
+                return Err("Attempt to decode a BitVec with too many bits".into())
+            }
+            // NOTE: Replace with `bits.div_ceil(Store::BITS)` if `int_roundings` is stabilised
+            let bytes_needed =
+                (bits / Store::BITS) as usize + (bits % Store::BITS != 0) as usize;
+
+            // NOTE: We could reduce allocations if it would be possible to directly
+            // decode from an `Input` type using a custom format (rather than default <u8, Lsb0>)
+            // for the `Bits` type.
+            let mut storage = Vec::with_capacity(bytes_needed);
+            input.read(&mut storage)?;
+
+            let decoder = scale_bits::decode_using_format_from(
+                &storage,
+                bit_format::<Store, Order>(),
+            )?;
+            let bits = decoder.collect::<Result<Vec<_>, _>>()?;
+            let bits = Bits::from_iter(bits);
+
+            Ok(DecodedBits(bits, PhantomData, PhantomData))
+        }
+    }
+
+    impl<Store: BitStore, Order: BitOrder> codec::Encode for DecodedBits<Store, Order> {
+        fn size_hint(&self) -> usize {
+            self.0.size_hint()
+        }
+
+        fn encoded_size(&self) -> usize {
+            self.0.encoded_size()
+        }
+
+        fn encode(&self) -> Vec<u8> {
+            scale_bits::encode_using_format(self.0.iter(), bit_format::<Store, Order>())
+        }
+    }
+}
+
+pub use bits::{
+    BitOrder,
+    BitStore,
+    DecodedBits,
+    Lsb0,
+    Msb0,
+};
```

Comment:
Probably it'd be good to have a few tests which assert that the decoded/encoded output lines up vs the BitVec equivalents (it's fine including bitvec as a dev dependency anyway).

Maybe something like a function which:
- takes an iterator of bools
- turn them into the various types above + their bitvec equivalents
- SCALE encode those all; make sure they line up
- decode again and make sure you get the expected output back

And then call that for some different bool arrays?

Reply:
Good call with tests! Turns out I had a bug or two... :sweat_smile: Added those in https://github.com/paritytech/subxt/pull/718/commits/f973fbc9e674e8816092e14ef094d82117f9044f